src/plugins/mf_priority.cpp

```diff
@@ -0,0 +1,230 @@
+
+std::map<int, std::map<std::string, double>> users;
+
+/******************************************************************************
+ *                                                                            *
+ *                           Helper Functions                                 *
+ *                                                                            *
+ *****************************************************************************/
+
+/*
+ * Calculate a user's job priority using the following factors:
+ *
+ * fairshare: the ratio between the amount of resources allocated vs. resources
+ *     consumed.
+ * urgency: a user-controlled factor to prioritize their own jobs.
+ */
+int64_t priority_calculation (flux_plugin_t *p,
+                              flux_plugin_arg_t *args,
+                              int userid,
+                              char *bank,
+                              int urgency)
+{
+    double fshare_factor = 0.0, priority = 0.0;
+    int fshare_weight;
+
+    std::map<int, std::map<std::string, double>>::iterator it;
+    std::map<std::string, double>::iterator inner_it;
+
+    fshare_weight = 100000;
+
+    if (urgency == FLUX_JOB_URGENCY_HOLD)
+        return FLUX_JOB_PRIORITY_MIN;
+
+    if (urgency == FLUX_JOB_URGENCY_EXPEDITE)
+        return FLUX_JOB_PRIORITY_MAX;
+
+    // search element in map of maps by key
+    it = users.find (userid);
+
+    // search for correct fshare value using passed-in bank; otherwise, use
+    // a default bank
+    if (bank != NULL) {
+        inner_it = it->second.find (bank);
+        fshare_factor = inner_it->second;
+    }
+    else {
+        inner_it = it->second.find ("default");
+        fshare_factor = inner_it->second;
+    }
+
+    priority = (fshare_weight * fshare_factor) + (urgency - 16);
+
+    return abs (round (priority));
+}
+
+
+/******************************************************************************
+ *                                                                            *
+ *                               Callbacks                                    *
+ *                                                                            *
+ *****************************************************************************/
+
+/*
+ * Unpack a payload from an external bulk update service and place it in the
+ * multimap datastructure.
+ */
+static void rec_update_cb (flux_t *h,
+                           flux_msg_handler_t *mh,
+                           const flux_msg_t *msg,
+                           void *arg)
+{
+    flux_plugin_t *p;
+    assert (sizeof arg == sizeof p);
+    memcpy (&p, &arg, sizeof arg);
+
+    char *uid, *fshare, *bank;
+
+    if (flux_request_unpack (msg, NULL, "{s:s, s:s, s:s}",
+                             "userid", &uid,
+                             "bank", &bank,
+                             "fairshare", &fshare) < 0) {
+        flux_log_error (h, "failed to unpack custom_priority.trigger msg");
```

Comment:
It would probably be good to add a test for this error condition. Just send an empty payload or other invalid input to the mf_priority plugin's service and ensure that the RPC fails expectedly, and that the job manager or instance doesn't crash ;-)

Reply:
Does sending an empty or invalid input payload causes an error to happen just before it gets to the `flux_log_error` message? For example, if an empty payload is sent:

```
Traceback (most recent call last):
  File "bad_payload.py", line 8, in <module>
    flux.Flux().rpc("job-manager.mf_priority.rec_update", {}).get()
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 78, in get
    resp_str = self.get_str()
  File "/usr/lib64/flux/python3.6/flux/util.py", line 76, in func_wrapper
    retval = func(future, *args, **kwargs)
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 72, in get_str
    self.pimpl.flux_rpc_get(payload_str)
  File "/usr/lib64/flux/python3.6/flux/util.py", line 52, in func_wrapper
    raise EnvironmentError(error.errno, errmsg.decode("utf-8")) from None
OSError: [Errno 71] Object item not found: userid
```

Or if an invalid payload gets sent:

```
Traceback (most recent call last):
  File "bad_payload.py", line 8, in <module>
    flux.Flux().rpc("job-manager.mf_priority.rec_update", data).get()
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 78, in get
    resp_str = self.get_str()
  File "/usr/lib64/flux/python3.6/flux/util.py", line 76, in func_wrapper
    retval = func(future, *args, **kwargs)
  File "/usr/lib64/flux/python3.6/flux/rpc.py", line 72, in get_str
    self.pimpl.flux_rpc_get(payload_str)
  File "/usr/lib64/flux/python3.6/flux/util.py", line 52, in func_wrapper
    raise EnvironmentError(error.errno, errmsg.decode("utf-8")) from None
OSError: [Errno 71] Expected string, got integer
```

I could very well be doing something wrong though. Is there a proper way to handle this failed unpack when a payload is empty or of invalid format?

Reply:
That looks like the expected result to me, since you send the result of `flux_msg_last_error()` in the error response. You may have to run `flux dmesg` to see the logged error.

Reply:
Ah, okay, thanks for pointing that out. I haven't used `flux dmesg` before, so apologies if I am totally using it wrong. I just pushed a `--fixup` commit (6c27d7a) that adds a test that uses `flux dmesg` to make sure that `flux_log_error` reports the right message.

Reply:
LGTM. I think you are using it correctly. In this case, you are getting an error back from the RPC which indicates the incorrect payload was caught by your plugin, so IMO it would be sufficient to ensure the `bad_payload.py` script fails with an expected error message.